src/dfx/src/lib/operations/canister/create_canister.rs

```diff
@@ -85,6 +87,59 @@ pub async fn create_canister(
         return reserve_canister_with_playground(env, canister_name).await;
     }
 
+    // Specified ID from the command line takes precedence over the one in dfx.json.
+    let specified_id = match (
+        config_interface.get_specified_id(canister_name)?,
+        specified_id_from_cli,
+    ) {
+        (Some(specified_id_from_json), Some(specified_id_from_cli)) => {
+            if specified_id_from_json != specified_id_from_cli {
+                warn!(
+                    env.get_logger(),
+                    "Canister '{0}' has a specified ID in dfx.json: {1},
+which is different from the one specified in the command line: {2}.
+The command line value will be used.",
+                    canister_name,
+                    specified_id_from_json,
+                    specified_id_from_cli
+                );
+            }
+            Some(specified_id_from_cli)
+        }
+        (Some(specified_id_from_json), None) => Some(specified_id_from_json),
+        (None, Some(specified_id_from_cli)) => Some(specified_id_from_cli),
+        (None, None) => None,
+    };
+
+    // Replace call_sender with wallet canister when necessary.
+    let call_sender =
+        if specified_id.is_none() && !no_wallet && !matches!(call_sender, CallSender::Wallet(_)) {
```

Comment:
This expression has a lot of "not" in it.  I think we can rework it to be a little easier to follow, something like this (but please double-check the boolean logic):

```
if specified_id.is_some() || no_wallet || matches!(call_sender, Wallet...) {
    *call_sender
} else {
    match ...
}

```